src/info/sigma/zcoinEngineExtension.js

```diff
@@ -41,8 +52,57 @@ export class ZcoinEngineExtension implements CurrencyEngineExtension {
     this.runLooperIfNeed()
   }
 
+  onBalanceChanged() {
+    logger.info('zcoinEngineExtension -> onBalanceChanged called')
+
+    this.currencyEngine.callbacks.onBalanceChanged(
+      this.currencyEngine.currencyCode,
+      this.engineState.getBalance({ mintedBalance: true })
+    )
+  }
+
+  async saveTx(edgeTransaction: EdgeTransaction) {
+    logger.info('zcoinEngineExtension -> saveTx called')
+
+    const { otherParams = {} } = edgeTransaction
+    const { mintsForSave = [] } = otherParams
+
+    await this.zcoinStateExtensions.appendMintedCoins(mintsForSave)
+  }
+
   async loop() {
     logger.info('zcoinEngineExtension -> loop called')
+
+    // TODO: can move into newTx
+    const utxos = this.engineState.getUTXOs()
+    const needToMint = sumUtxos(utxos)
+    logger.info('zcoinEngineExtension -> Not minted balance: ' + needToMint)
+    let needToMintStr = needToMint.toString()
+    if (bns.gt(needToMintStr, denominations[0])) {
+      if (
+        bns.mul(bns.div(needToMintStr, denominations[0]), denominations[0]) ===
+        needToMintStr
+      ) {
+        // can't mint all balance, because of fee
+        needToMintStr = bns.sub(needToMintStr, denominations[0])
+      }
+      logger.info('zcoinEngineExtension -> Trying to mint: ' + needToMintStr)
+      const edgeInfo: EdgeSpendInfo = {
+        currencyCode: 'XZC',
+        spendTargets: [
+          {
+            publicAddress: this.keyManager.getChangeAddress(),
+            nativeAmount: needToMintStr
+          }
+        ]
+      }
+      // mint and getMintMetadataLoop must not work paralelly in order to avoid problems related to modifying mint.json parallely
+      await this.mint(edgeInfo)
+    }
+
+    await this.getMintMetadataLoop()
+
+    this.onBalanceChanged()
```

Comment:
check if we need to call onBalanceChanged, then actually call

Reply:
done